### ISSUE TITLE
Request for help.

### ISSUE THREAD

**Issue by DxsSucuk** [deepjavalibrary/djl#1665]
# Information
Hey there!

I wanted to ask for help since I couldn't figure out how to fix a certain bug, and because I did not get an answer in 5 days on the DJL Slack channel.

Related [Slack](https://app.slack.com/client/TPX8YGQTW/C01AURG857U/thread/C01AURG857U-1652044197.694489)
Related [Exception](https://pastebin.com/8ksTbbh1)

# Issue 
The Issue that I am having is that the BinaryImageTranslator I created has Issue with any Image given to it, I could not figure out why exactly but as of it looks there is an Issue with the resizing/reshaping of the Input Image.

**Comment by frankfliu** [comment]
@DxsSucuk 

Sorry for the delay. We were busy in 0.17.0 release.
@zachgk can you follow up on this issue?

**Comment by DxsSucuk** [comment]
All good I understand, expecially on a project like this which Im pretty sure takes a lot of "big brain".

**Comment by lanking520** [comment]
Hi @DxsSucuk if you don't find, can you send the code and model (if it is open sourced) to us and we will take a look. And also feel free to grab me (Lanking) a slack call, happy to hop in and help :)

**Comment by DxsSucuk** [comment]
Sure thing!
The source is on my Git Account [here](https://github.com/DxsSucuk/PissAI)

**Comment by DxsSucuk** [comment]
Any Idea of what can be broke?

**Comment by lanking520** [comment]
@DxsSucuk here is the solution:

In line: https://github.com/DxsSucuk/PissAI/blob/master/src/main/java/de/presti/pissai/trainer/BinaryImageTranslator.java#L24-L32

Change them to:

```
    @Override
    public Batchifier getBatchifier() {
        return null;
    }

    @Override
    public Float processOutput(TranslatorContext ctx, NDList list) {
        return list.singletonOrThrow().toFloatArray()[0];
    }
```

To pass the run. The reason why the error happened is:

1. You use StackBatchifier, where it tried to unbatchify at the beginning, and you don't need this right now.

2. The output is a NDArray with Shape() empty, there are single value inside. So you should try to get Float array and extrat the first value from it.

**Comment by DxsSucuk** [comment]
Thanks @lanking520 !

That fixed the Issue, but now it returns a Number that is quite weird. The Number is -0.727, and that is quite confusing.

**Comment by DxsSucuk** [comment]
@lanking520 Any Idea why that could be?

**Comment by lanking520** [comment]
In my mind it can caused by you haven't apply a softmax function to make the result fall from 0 -1. What is the expected number for you? @zachgk

**Comment by DxsSucuk** [comment]
> In my mind it can caused by you haven't apply a softmax function to make the result fall from 0 -1. What is the expected number for you? @zachgk 

For me it would be a numer between 0 and 1 like an precentage. Means 0.7 = 70% similiar.

**Comment by DxsSucuk** [comment]
@lanking520 Just tried to use softmax, to fix it but now it always return 1.0

**Comment by zachgk** [comment]
You want sigmoid, not softmax. Softmax is for multi-class while sigmoid is for binary. Also, the sigmoid is automatically applied by the `SigmoidBinaryCrossEntropyLoss`, so you don't need to add it to your model. You can just think of it as the `sigmoid(modelOutput)` would be the probability you are looking for.

You can also make sense of the pre-sigmoid result too. Maybe look at [the graph on wikipedia](https://en.wikipedia.org/wiki/Sigmoid_function). More positive means higher probability, more negative means lower probability, and 0 means 50-50.

Or, `sigmoid(-7.2680035) = 0.0006970170 = .06970170%`

**Comment by DxsSucuk** [comment]
Is there a math equation to "convert" it into a percentage.

**Comment by zachgk** [comment]
Yeah. You can use the DJL operator `Activation.sigmoid(array)`. The actual equation for sigmoid is $sigmoid(x) = \frac{1}{1 + e^{-x}}$. These give you the decimal form and then you can multiply by 100 to get the percentage if you prefer that

**Comment by DxsSucuk** [comment]
Thank you! It really help, now the only thing is that it says that it is only abou 6% similiar, which it shouldn't since I used an Image it was trained with.

**Comment by zachgk** [comment]
What are you getting for your training/validation accuracy? Are they increasing as you train?

**Comment by DxsSucuk** [comment]
I did not setup anything to show me that information, as far as I know.

**Comment by zachgk** [comment]
You have the logging training listeners (https://github.com/DxsSucuk/PissAI/blob/master/src/main/java/de/presti/pissai/trainer/ModelTrainer.java#L114), so it should be printing the loss and accuracy to the stdout

**Comment by DxsSucuk** [comment]
It actually does not, maybe a Logger Configuration Issue?

**Comment by DxsSucuk** [comment]
I checked and tried using a LogBack Configuration that I use on another project to store debug data in another File, but I can't seem to find anything @zachgk

**Comment by DxsSucuk** [comment]
After reading alot of the documentation, I found out that the Example class for MNIST already does what is needed, so I checked it out and [this](https://pastebin.com/dppGsKJc) should be the result you are searching for right? @zachgk

**Comment by DxsSucuk** [comment]
But the Model is trained to be binairy? Wouldn't it be the currect way to let it only know how dream looks like, so it can determin if a new Image suits the known pattern or not?

**Comment by DxsSucuk** [comment]
And I just checked again with the same Image and got [this](https://paste.gg/p/anonymous/d89e7fed1cbd401a8b8edaf3106bec3f), either my percentage calculation is wrong or something else in my code. And yeah I used the exact same Image I used as training Image. When I try the same, but with the Model that has 41 epochs, I get 3%.

**Comment by zachgk** [comment]
Think about it like this. The model is trying to determine rules that help explain what we are asking of it. And, it could learn the rule `isDream(image) = true`. The rule is easy to learn and based on the dataset it is 100% accurate. So there's no reason not to do it.

That's why we need negative examples. Then, it would learn that the always true rule is not very good and it would work to come up with better rules.

**Comment by DxsSucuk** [comment]
Aah, basically a way to prevent the Model from just always saying "Yeah it is dream", since that would be true on the training dataset. How would I fix this exactly? Just add another folder into the Imagefolder with alot of Images of different stuff? Or?

**Comment by zachgk** [comment]
Yeah, adding another folder with a variety of random stuff would work. Especially consider examples that you specifically want it to not misclassify

**Comment by DxsSucuk** [comment]
Any Idea? @zachgk ?

**Comment by zachgk** [comment]
The only change you made was adding new images? Well, it could be that the images are being loaded differently from the original images. It could be different sizes, but that shouldn't be the problem because you have a resize in your image loading pipeline. Is it possible you have added some greyscale images that are being loaded as [h,w] instead of [h,w,3] or some png images being loaded with an alpha channel?

Is it failing on the first batch or is it making it through some number of batches before failing?

**Comment by DxsSucuk** [comment]
It is failing instantly. And I still get the same exact error, I check for Images with more or less then 3 channels by making a small app:




            try {
                BufferedImage bufferedImage = ImageIO.read(file);
                if (bufferedImage == null) {
                    file.delete();
                    System.out.println("Deleted " + file.getName() + ", cause null");
                    continue;
                }

                if (bufferedImage.getWidth() < 256 || bufferedImage.getHeight() < 256) {
                    file.delete();
                    System.out.println("Deleted " + file.getName() + " Width: " + bufferedImage.getWidth() + ", Height: " + bufferedImage.getHeight());
                } else {
                    Raster ras = bufferedImage.getRaster();

                    if (ras.getNumDataElements() != 3) {
                        file.delete();
                        System.out.println("Deleted " + file.getName() + " cause its not a 3 channel image instead its a " + ras.getNumDataElements() + " channel image");
                    }
                }
            } catch (Exception exception) {
                file.delete();
                System.out.println("Deleted " + file.getName());
                exception.printStackTrace();
            }
            
this could detected about 50 Images and deleted them, but it still does not work.

**Comment by DxsSucuk** [comment]
How may I check this? @zachgk

**Comment by DxsSucuk** [comment]
How may I check this? @zachgk

**Comment by DxsSucuk** [comment]
So I couldn't find anything that could case the Issue, rather then it all being 3. But it worked before hand? Why does it not work now? It overally is quite confusing any Idea what I am doing wrong @zachgk ?

**Comment by DxsSucuk** [comment]
Any Ideas @zachgk ?

**Comment by zachgk** [comment]
Did you check the labels? Just try runnning something like:

```java
for (Batch batch : dataset.getData(manager) {
  System.out.println(batch.getLabels().singletonOrThrow());
}
```

**Comment by DxsSucuk** [comment]
This is the output I get. @zachgk 
```Batch 1 ND: (32) cpu() int64
Batch 2 ND: (32) cpu() int64
Batch 3 ND: (32) cpu() int64
Batch 4 ND: (32) cpu() int64
Batch 5 ND: (32) cpu() int64
Batch 6 ND: (13) cpu() int64

**Comment by DxsSucuk** [comment]
Well I did both just in case, and while debugging it shows me that there are two labels, as wanted. And the getSynset method gives me two labels aswell @zachgk .

**Comment by DxsSucuk** [comment]
Any Ideas @zachgk ?

**Comment by zachgk** [comment]
Ok. So, I checked out the code and got it running. Here is what you need to change:

Change [ModelTrainer.java:141](https://github.com/DxsSucuk/PissAI/blob/master/src/main/java/de/presti/pissai/trainer/ModelTrainer.java#L141) to

```java
sequentialBlock.addSingleton(a -> a.get(":, 1"));
```

The model returns a 2 class multi-class classification and this converts into a single class classification by turning it basically into class `1` vs not class `1`.

Change [ModelTrainer.java:156](https://github.com/DxsSucuk/PissAI/blob/master/src/main/java/de/presti/pissai/trainer/ModelTrainer.java#L156) to

```java
trainer.initialize(new Shape(32, 256 * 256 * 3));
```

You need to initialize it similarly to how the model is trained. Because we train it with batches, it should also be initialized with batches.

Finally, you probably want to reduce the default number of epochs. An epoch is how many times each data element is trained with, so it doesn't need to increase with the dataset size. Basically, just see how long it takes for the model to stop improving. When I was testing it, it seemed to be fully trained after 4 epochs

**Comment by DxsSucuk** [comment]
I am more then grateful for the help! I will check it out and check if it works!

**Comment by DxsSucuk** [comment]
Thank you! It fixes the Issue and lets me train the model and test it!
But I am unsure why it gives me a value of 67% confidence on a Image that isn't right.
It also gives me a 51% on a Image that is right. Any Ideas why?

**Comment by DxsSucuk** [comment]
Here are the [logs](https://paste.gg/p/anonymous/1f4df3b5b42f400ca388a3305b443e46)

**Comment by DxsSucuk** [comment]
After a bit of testing and playing around, I found out if I take away the float coming out of the prediction from 1 I will get a more realistic number.

This means on the Invalid Image I get the float value 0.6664029 and if I do 1 - 0.6664029 I get 0.3335971 which is about 33%.
And if I do the same on a valid Image I get a 71%. Unsure if this is the right way to do this, but it is a temporal solution.

**Comment by DxsSucuk** [comment]
Well after more testing, the temporal solution is not the right thing. I keeps false positives away, but there are alot of false negatives.

**Comment by DxsSucuk** [comment]
@zachgk ?

**Comment by DxsSucuk** [comment]
Still could not find anything that could help me except that I would need to get even more Images.
@zachgk ?

**Comment by zachgk** [comment]
When I mentioned the line `sequentialBlock.addSingleton(a -> a.get(":, 1"))`, it means that class `1` is treated as true and not `1` is treated as false. I could have gotten that backwards and class `0` should really be true.

But, you should expect the accuracy to be fairly limited. Deep Learning is connected to big data because it often takes massive datasets to train models to be accurate. One thing you may want to do is to move maybe `20%` of your data into a test set and use it to check how accurate your model is after it is trained

Another option you can take is to adopt transfer learning. This is when you take a model already trained on a large dataset and then retrain it on a newer dataset. It often takes some amount of understanding of images from the original dataset that can help it get better results on the newer dataset. If you want to try this, check out https://github.com/deepjavalibrary/djl/blob/master/examples/src/main/java/ai/djl/examples/training/transferlearning/TrainResnetWithCifar10.java for an example of what transfer learning looks like
